collector/src/etw_parser.rs

```diff
@@ -0,0 +1,677 @@
+
+Pmc events increase monotonically per CPU (core) and are cumulative. Their data is associated
+with the following CSwitch event (they will have the same timestamp). We need to keep track of
+when the rustc process has been scheduled onto a CPU (CSwitch to rustc.exe) and when it has been
+scheduled off of that same CPU and then calculate the difference of the PMC and add it to the
+running total.
+
+Algorithm:
+- Read in the header and determine which columns correspond to the values we care about
+- Validate the OS Version line does not have lost events or buffers
+- Locate all the events for the rustc.exe process (probably validate there is only one present) and parse them into useful data structures
+- Do basic sanity checking
+    - Validate we have 1 P-Start & 1 P-End at beginning and end of timeline
+    - Validate every Pmc has a next CSwitch event at the same timestamp
+- It may be convenient at this point to rewrite into a stream of (pmc data, CSwitch data)
+- Count up deltas of PMC data on each CPU -> this is the total for the counter
+- Return results
+*/
+
+#[derive(Debug, Eq, PartialEq)]
+struct EventHeader {
+    name: String,
+    columns: Vec<String>,
+}
+
+const PROCESS_START: &str = "P-Start";
+const PROCESS_END: &str = "P-End";
+const PMC: &str = "Pmc";
+const CSWITCH: &str = "CSwitch";
+const TIMESTAMP: &str = "TimeStamp";
+const PROCESS_NAME_PID: &str = "Process Name ( PID)";
+const THREAD_ID: &str = "ThreadID";
+const INSTRUCTION_RETIRED: &str = "InstructionRetired";
+const TOTAL_CYCLES: &str = "TotalCycles";
+const NEW_PROCESS_NAME_PID: &str = "New Process Name ( PID)";
+const OLD_PROCESS_NAME_PID: &str = "Old Process Name ( PID)";
+const CPU: &str = "CPU";
+const PARENT_PID: &str = "ParentPID";
+
+fn parse_header(r: &mut dyn BufRead) -> anyhow::Result<Vec<EventHeader>> {
+    let mut line = String::new();
+    r.read_line(&mut line)?;
+
+    anyhow::ensure!(line.trim() == "BeginHeader", "BeginHeader line not found at beginning of file");
+
+    let mut headers = Vec::with_capacity(6);
+
+    while { line.clear(); r.read_line(&mut line)? != 0 } {
+        let l = line.trim();
+        if l == "EndHeader" {
+            break;
+        }
+
+        let mut contents = l.split(',');
+        let name = contents.next().expect("event header was missing name").trim();
+
+        if name == PROCESS_START || name == PROCESS_END || name == PMC || name == CSWITCH {
+            headers.push(EventHeader {
+                name: name.to_string(),
+                columns: contents.map(|s| s.trim().to_string()).collect(),
+            });
+        }
+    }
+
+    anyhow::ensure!(line.trim() == "EndHeader", "EndHeader line not found");
+
+    Ok(headers)
+}
+
+fn validate_os_header_line(r: &mut dyn BufRead) -> anyhow::Result<()> {
+    let mut line = String::new();
+    r.read_line(&mut line)?;
+
+    anyhow::ensure!(line.starts_with("OS Version"), "OS version line not found");
+
+    let components: Vec<_> = line.split(',').collect();
```

Comment:
Nit: is allocating a vec here necessary?

Reply:
Probably not but I didn't want to resplit the string twice. I'm not sure which is more expensive...